src/graph/sampling/get_node_types_cpu.cc

```diff
@@ -0,0 +1,67 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file graph/sampling/get_node_types_cpu.cc
+ * \brief DGL sampler - CPU implementation of random walks with OpenMP
+ */
+
+#include <dgl/array.h>
+#include <dgl/base_heterograph.h>
+#include <utility>
+#include "randomwalks.h"
+
+namespace dgl {
+
+using namespace dgl::runtime;
+using namespace dgl::aten;
+
+namespace sampling {
+
+namespace impl {
+
+template<DLDeviceType XPU, typename IdxType>
+TypeArray GetNodeTypesFromMetapath(
+    const HeteroGraphPtr hg,
+    const TypeArray metapath) {
+  uint64_t num_etypes = metapath->shape[0];
+  TypeArray result = TypeArray::Empty(
+      {metapath->shape[0] + 1}, metapath->dtype, metapath->ctx);
+
+  const IdxType *metapath_data = static_cast<IdxType *>(metapath->data);
+  IdxType *result_data = static_cast<IdxType *>(result->data);
+
+  dgl_type_t etype = metapath_data[0];
+  dgl_type_t srctype = hg->GetEndpointTypes(etype).first;
+  dgl_type_t curr_type = srctype;
+  result_data[0] = curr_type;
+
+  for (uint64_t i = 0; i < num_etypes; ++i) {
+    etype = metapath_data[i];
+    auto src_dst_type = hg->GetEndpointTypes(etype);
+    dgl_type_t srctype = src_dst_type.first;
+    dgl_type_t dsttype = src_dst_type.second;
```

Comment:
Use local variable for `etype`. `srctype` shadows another variable outside of the loop. I will shorten L32-L34 as:
```
dgl_type_t curr_type = hg->GetEndpointTypes(metapath_data[0]).first;
```